library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/VanillaRecipeBuilders.java

```diff
@@ -0,0 +1,407 @@
+		 * @param ingredient the ingredient
+		 * @return this builder
+		 */
+		public ShapedRecipeBuilder ingredient(char key, Ingredient ingredient) {
+			boolean success = false;
+
+			for (String line : pattern) {
+				for (int i = 0; i < line.length(); i++) {
+					char c = line.charAt(i);
+
+					if (c == key) {
+						this.ingredients.put(key, ingredient);
+
+						success = true;
+						break;
+					}
+				}
+
+				if (success) break;
+			}
+
+			return this;
+		}
+
+		/**
+		 * Puts the specified items as the accepted ingredient at the specified key.
+		 *
+		 * @param key   the key of the ingredient
+		 * @param items the items as ingredient
+		 * @return this builder
+		 * @see #ingredient(char, Ingredient)
+		 */
+		public ShapedRecipeBuilder ingredient(char key, ItemConvertible... items) {
+			return this.ingredient(key, Ingredient.ofItems(items));
+		}
+
+		/**
+		 * Puts the specified item tag as the accepted ingredient at the specified key.
+		 *
+		 * @param key the key of the ingredient
+		 * @param tag the item tag as ingredient
+		 * @return this builder
+		 * @see #ingredient(char, Ingredient)
+		 */
+		public ShapedRecipeBuilder ingredient(char key, TagKey<Item> tag) {
+			return this.ingredient(key, Ingredient.method_8106(tag));
+		}
+
+		/**
+		 * Puts the specified item stacks as the accepted ingredient at the specified key.
+		 *
+		 * @param key    the key of the ingredient
+		 * @param stacks the item stacks as ingredient
+		 * @return this builder
+		 * @see #ingredient(char, Ingredient)
+		 */
+		public ShapedRecipeBuilder ingredient(char key, ItemStack... stacks) {
+			return this.ingredient(key, Ingredient.ofStacks(stacks));
+		}
+
+		/**
+		 * Builds the shaped crafting recipe.
+		 *
+		 * @param id    the identifier of the recipe
+		 * @param group the group of the recipe
+		 * @return the shaped recipe
+		 */
+		@Override
+		public ShapedRecipe build(Identifier id, String group) {
+			super.build(id, group);
+			DefaultedList<Ingredient> ingredients = getIngredients(this.pattern, this.ingredients, this.width, this.height);
+			return new ShapedRecipe(id, group, this.width, this.height, ingredients, this.output);
+		}
+	}
+
+	public static final class ShapelessRecipeBuilder extends Builder<ShapelessRecipeBuilder, ShapelessRecipe> {
+		private final Set<Ingredient> ingredients = new HashSet<>();
+
+		public ShapelessRecipeBuilder(ItemStack output) {
+			this.output = output;
+		}
+
+		/**
+		 * Adds an ingredient.
+		 *
+		 * @param ingredient the ingredient
+		 * @return this builder
+		 */
+		public ShapelessRecipeBuilder ingredient(Ingredient ingredient) {
+			this.ingredients.add(ingredient);
+			return this;
+		}
+
+		/**
+		 * Puts the specified items as the accepted ingredient at the specified key.
+		 *
+		 * @param items the items as ingredient
+		 * @return this builder
+		 * @see #ingredient(Ingredient)
+		 */
+		public ShapelessRecipeBuilder ingredient(ItemConvertible... items) {
```

Comment:
For both this and the `ItemStack` version, do all the parameters count as one ingredient and are interchangeable, or are they all separate? I would clarify the javadocs on this

Reply:
> `Adds item stacks as an ingredient.`
for the item stack version

I thought it was clear that it would be one ingredient?